### PR TITLE
Refactors `settings::RuleStrategy.rule` as `Rule` enum

### PR DESCRIPTION
**Pull request by geeksambhu**
## Description

Fix https://github.com/kubewarden/user-group-psp-policy/issues/16

## Test

To test this pull request, you can run the following commands:

`make test`

All test should pass

### CODE REVIEW

**Review by jvanz — CHANGES_REQUESTED** [review]
That's good starting point Shiva! However, I have some change requests.

In the Rust code you can remove the `ranges` field in the `RuleStrategy` struct. Now the ranges will be inside the `Rule` enum. Making the `ranges` field obsolete. This means that you can also simplify the code where the validations are performed. 

Once you remove the `ranges` field, another problem comes up.  The settings will serialize as the following:

```json
{
  "run_as_user": {
    "rule": {
      "MustRunAs": [
        {
          "min": 1000,
          "max": 2000
        }
      ]
    },
    "overwrite": true
  },
  "run_as_group": {
    "rule": "RunAsAny",
    "overwrite": false
  },
  "supplemental_groups": {
    "rule": "RunAsAny",
    "overwrite": false
  }
}
```

Now, take a look how the stable version is serialized (you can see more examples in the README):

```json
{
  "run_as_user": {
    "rule": "RunAsAny"
  },
  "run_as_group": {
    "rule": "MayRunAs",
    "ranges": [
      {
        "min": 1000,
        "max": 2000
      },
      {
        "min": 2001,
        "max": 3000
      }
    ]
  },
  "supplemental_groups":{
    "rule": "MayRunAs",
    "ranges": [
      {
        "min": 1000,
        "max": 2000
      },
      {
        "min": 2001,
        "max": 3000
      }
    ]
  }
}
```

Note this will require a major version update. I'm not sure if this should be the case. Therefore, I would say to research if you can get the same JSON as in the stable version changing serde configs or customize the way serde serialize/deserialize the object.  See the serde [docs](https://serde.rs/custom-serialization.html) to see how to do that.

Note, the serialization/deserialization customization may not require the change in the internal struct representation. If you think this will be required, let us know.

**Comment by viccuad** [comment]
After consideration, I think that doing the checks via the deserializer makes everything needlessly complicated. The card did indeed ask to try to do it via de Deserializer, which I know think it's a wrong approach. I'm sorry for that.

I would be happy just having

```
pub(crate) struct RuleStrategy {
    pub rule: Rule,
    pub ranges: Vec<IDRange>,
    pub overwrite: bool,
}
```

Where `Rule` is an enum with variants `MustRunAs`, `RunAsAny`, `MustRunAsNonRoot`, `MayRunAs` (with the correct serde attributes for it to deserialize correctly). And having the validation of the settings being performed on runtime at `validate()`.

File: src/settings.rs [hunk 1]
```diff
@@ -255,23 +267,23 @@ mod tests {
...
+                    rule: Rule::RunAsAny,
                     ranges: vec![IDRange {
                         min: 1000,
                         max: 1010,
                     }],
                     ..Default::default()
                 },
                 supplemental_groups: RuleStrategy {
-                    rule: String::from(rule),
+                    rule: Rule::MustRunAs,
```
Comment (viccuad) [inline src/settings.rs:286]: I think this is incorrect, shouldn't it be `rule: rule`, as in line 237?

On line 226, we iterate through all allowed rules values, and we create to possible settings, one where we change the `run_as_group`, and one where we change the `supplemental_groups`.

File: src/settings.rs [hunk 2]
```diff
@@ -11,10 +11,22 @@ pub(crate) struct IDRange {
     pub max: i64,
 }
 
+#[derive(Serialize, Deserialize, Debug)]
+pub(crate) enum Rule {
+    #[serde(rename = "MustRunAs")]
```
Comment (viccuad) [inline src/settings.rs:16]: we can drop all serde renames, they aren't needed, they set the same style as the enum variants already have.

**Review by viccuad — APPROVED** [review]
LGTM!

**Comment by viccuad** [comment]
tagged in main as `v0.4.8`, will be released shortly.

**Pull request merged**